### PR TITLE
[Improve][Zeta] Move SaveMode behavior to master

### PR DESCRIPTION
**Pull request by Hisoka-X**
### Purpose of this pull request
This PR move savemode logic into master node.
Mainly for two purposes:
1. Maintain the same behavior as the Rest submit job. Before this, the SaveMode logic of Rest mode was executed on the master, but the client mode was executed on the client side.
2. Prepare for related SaveMode event processing. Based on purpose 1, we can ensure consistent behavior of SaveMode related events.

### Does this PR introduce _any_ user-facing change?
no

### How was this patch tested?
add new test.

### Check list

* [ ] If any new Jar binary package adding in your PR, please add License Notice according
  [New License Guide](https://github.com/apache/seatunnel/blob/dev/docs/en/contribution/new-license.md)
* [ ] If necessary, please update the documentation to describe the new feature. https://github.com/apache/seatunnel/tree/dev/docs
* [ ] If you are contributing the connector code, please check that the following files are updated:
  1. Update change log that in connector document. For more details you can refer to [connector-v2](https://github.com/apache/seatunnel/tree/dev/docs/en/connector-v2)
  2. Update [plugin-mapping.properties](https://github.com/apache/seatunnel/blob/dev/plugin-mapping.properties) and add new connector information in it
  3. Update the pom file of [seatunnel-dist](https://github.com/apache/seatunnel/blob/dev/seatunnel-dist/pom.xml)
* [ ] Update the [`release-note`](https://github.com/apache/seatunnel/blob/dev/release-note.md).

### CODE REVIEW

File: seatunnel-engine/seatunnel-engine-core/src/main/java/org/apache/seatunnel/engine/core/dag/logical/LogicalDag.java [hunk 1]
```diff
@@ -129,40 +141,47 @@ public int getClassId() {
...
         }
 
         out.writeInt(edges.size());
-
         for (LogicalEdge edge : edges) {
             out.writeObject(edge);
         }
 
+        out.writeInt(saveModeList.size());
+        for (SupportSaveMode saveMode : saveModeList) {
```
Comment (dailai) [inline seatunnel-engine/seatunnel-engine-core/src/main/java/org/apache/seatunnel/engine/core/dag/logical/LogicalDag.java]: Should we consider the DataSerializable of this?

Comment (Hisoka-X) [reply seatunnel-engine/seatunnel-engine-core/src/main/java/org/apache/seatunnel/engine/core/dag/logical/LogicalDag.java]: In fact, the current SupportSaveMode declarations are all on the SeaTunnelSink implementation class, and all SeaTunnelSinks are serializable by default. It is also engine-independent.

File: seatunnel-api/src/main/java/org/apache/seatunnel/api/sink/SaveModeExecuteLocation.java [hunk 2]
```diff
@@ -0,0 +1,23 @@
...
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.seatunnel.api.sink;
+
+public enum SaveModeExecuteLocation {
+    CLIENT,
```
Comment (hailin0) [inline seatunnel-api/src/main/java/org/apache/seatunnel/api/sink/SaveModeExecuteLocation.java:22]: Mark it as obsolete for possible future removal？

File: docs/en/concept/JobEnvConfig.md [hunk 3]
```diff
@@ -33,6 +33,10 @@ This parameter configures the parallelism of source and sink.
 
 Used to control the default retry times when a job fails. The default value is 3, and it only works in the Zeta engine.
 
+### savemode.execute.location
+
+This parameter is used to specify the location of the savemode when the job is executed in the Zeta engine. The default value is `CLUSTER`, which means that the savemode is executed on the cluster. If you want to execute the savemode on the client, you can set it to `CLIENT`.
```
Comment (hailin0) [inline docs/en/concept/JobEnvConfig.md]: `CLIENT`

Mark it as obsolete for possible future removal？

**Review by hailin0 — COMMENTED** [review]
LGTM

File: seatunnel-engine/seatunnel-engine-server/src/main/java/org/apache/seatunnel/engine/server/master/JobMaster.java [hunk 4]
```diff
@@ -333,6 +361,30 @@ public void run() {
...
+                    new SaveModeExecuteWrapper(handler).execute();
+                } catch (Exception e) {
+                    throw new SeaTunnelRuntimeException(HANDLE_SAVE_MODE_FAILED, e);
+                }
+            }
+        } else if (sink.getClass()
+                .getName()
+                .equals(
+                        "org.apache.seatunnel.connectors.seatunnel.common.multitablesink.MultiTableSink")) {
+            // TODO we should not use class name to judge the sink type
```
Comment (hailin0) [inline seatunnel-engine/seatunnel-engine-server/src/main/java/org/apache/seatunnel/engine/server/master/JobMaster.java:379]: Please explain the detailed reason. I see that it is possible to implement `SaveModeHandler`

Comment (Hisoka-X) [reply seatunnel-engine/seatunnel-engine-server/src/main/java/org/apache/seatunnel/engine/server/master/JobMaster.java:379]: It is difficult to perform some system-level operations at the connector level, such as create savemode-related events. I will solve this problem in the next PR.

File: docs/en/concept/JobEnvConfig.md (hunk 3 above)
Comment (Hisoka-X) [reply docs/en/concept/JobEnvConfig.md]: done

**Comment by hailin0** [comment]
Please fix ci error

**Comment by Hisoka-X** [comment]
> Please fix ci error

done

**Review by hailin0 — APPROVED** [review]
LGTM

**Review by EricJoy2048 — APPROVED** [review]

**Pull request merged**